File: 416-PartitionEqualSubsetSum/416-PartitionEqualSubsetSum.py
# Last updated: 23/03/2026, 20:41:54
1class Solution:
2    def canPartition(self, nums: List[int]) -> bool:
3        memo = {}
4        target_sum = sum(nums) // 2 
5
6        if sum(nums) % 2 == 1: 
7            return False
8
9
10        def dp(i,sm,target_sum): 
11            
12            if (i,sm) in memo: 
13                return memo[(i,sm)]
14
15            if sm == target_sum: 
16                return True
17            
18            if i >= len(nums): 
19                return False
20
21            memo[(i,sm)] = dp(i+1,sm,target_sum) or dp(i+1,sm+nums[i],target_sum)
22            return memo[(i,sm)]
23
24        return dp(0,0,target_sum)

File: 416-PartitionEqualSubsetSum/416-PartitionEqualSubsetSum.cpp
// Last updated: 23/03/2026, 20:42:37
1class Solution {
2public:
3    bool canPartition(vector<int>& nums) {
4        int total = accumulate(nums.begin(), nums.end(), 0);
5        if (total % 2 == 1) return false;
6
7        int target = total / 2;
8        int n = nums.size();
9        vector<vector<int>> memo(n, vector<int>(target + 1, -1));
10
11        function<bool(int, int)> dp = [&](int i, int sm) -> bool {
12            if (sm == target) return true;
13            if (i >= n || sm > target) return false;
14            if (memo[i][sm] != -1) return memo[i][sm];
15
16            memo[i][sm] = dp(i + 1, sm) || dp(i + 1, sm + nums[i]);
17            return memo[i][sm];
18        };
19
20        return dp(0, 0);
21    }
22};